lib/webdriveragent.js

```diff
@@ -190,21 +190,53 @@ class WebDriverAgent {
    *   }
    * }
    *
-   * @return {Promise<any?>} State Object
+   * @param {number} [timeoutMs=0] If the given timeoutMs is zero or negative number,
+   *                               this function will return the response of `/status` immediately. If the given timeoutMs,
+   *                               this function will try to get the response of `/status` up to the timeoutMs.
+   * @return {Promise<import('@appium/types').StringRecord|null>} State Object
+   * @throws {Error} If there was an error within timeoutMs timeout.
+   *                 No error is raised if zero or negative number for the timeoutMs.
    */
-  async getStatus () {
+  async getStatus (timeoutMs = 0) {
     const noSessionProxy = new NoSessionProxy({
       server: this.url.hostname,
       port: this.url.port,
       base: this.basePath,
       timeout: 3000,
     });
+
+    const sendGetStatus = async () => await /** @type import('@appium/types').StringRecord */ (noSessionProxy.command('/status', 'GET'));
+
+    if (_.isNil(timeoutMs) || timeoutMs <= 0) {
+      try {
+        return await sendGetStatus();
+      } catch (err) {
+        this.log.debug(`WDA is not listening at '${this.url.href}'. Original error:: ${err.message}`);
+        return null;
+      }
+    }
+
+    let lastError = null;
+    let status = null;
     try {
-      return await noSessionProxy.command('/status', 'GET');
+      await waitForCondition(async () => {
+        try {
+          status = await sendGetStatus();
+          return true;
+        } catch (err) {
+          lastError = err;
+        }
+        return false;
+      }, {
+        waitMs: timeoutMs,
+        intervalMs: 300,
+      });
     } catch (err) {
-      this.log.debug(`WDA is not listening at '${this.url.href}'`);
-      return null;
+      this.log.debug(`Failed to get the status endpoint in ${timeoutMs} ms. ` +
```

Comment:
The last error by ${this.url.href} was ${lastError}. Original error:: ${err.message}.` -> `The last error while accessing ${this.url.href}: lastError.message`